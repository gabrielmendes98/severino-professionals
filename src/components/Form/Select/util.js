/* eslint-disable react/display-name */
/* eslint-disable react/prop-types */
import MenuItem from '@material-ui/core/MenuItem';

export const getOptions = (options, hasEmptyValue, emptyStateMessage) => {
  if (options.length) {
    return hasEmptyValue ? [{ label: '', value: '' }, ...options] : options;
  }
  return [
    {
      label: emptyStateMessage || 'Não há opções',
      value: '',
      disabled: true,
    },
  ];
};

export const renderOption = name => (props, index) => {
  const { label, value } = props;

  return (
    <MenuItem
      id={`${name}-option-${index}`}
      value={value}
      key={value}
      alt={label}
      {...props}
    >
      {label}
    </MenuItem>
  );
};
