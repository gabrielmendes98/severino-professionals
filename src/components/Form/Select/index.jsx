import { useMemo } from 'react';
import PropTypes from 'prop-types';
import MuiTextField from '@material-ui/core/TextField';
import { useField } from 'formik';
import { getOptions, renderOption } from './util';

const Select = ({
  name,
  options = [],
  value,
  hasEmptyValue,
  emptyStateMessage,
  SelectProps,
  id,
  ...otherProps
}) => {
  const [field, meta] = useField(name);

  const buildOptions = useMemo(
    () =>
      getOptions(options, hasEmptyValue, emptyStateMessage).map(
        renderOption(name),
      ),
    [name, hasEmptyValue, options, emptyStateMessage],
  );

  const selectProps = {
    displayEmpty: hasEmptyValue && Boolean(options.length),
    ...SelectProps,
  };

  const configSelect = {
    ...field,
    select: true,
    variant: 'outlined',
    fullWidth: true,
    size: 'small',
    id: id || name,
    ...otherProps,
  };

  if (meta && meta.touched && meta.error) {
    configSelect.error = true;
    configSelect.helperText = meta.error;
  }

  return (
    <MuiTextField {...configSelect} SelectProps={selectProps}>
      {buildOptions}
    </MuiTextField>
  );
};

Select.propTypes = {
  name: PropTypes.string.isRequired,
  id: PropTypes.string,
  options: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
        .isRequired,
    }),
  ),
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  hasEmptyValue: PropTypes.bool,
  emptyStateMessage: PropTypes.string,
  SelectProps: PropTypes.object,
};

export default Select;
