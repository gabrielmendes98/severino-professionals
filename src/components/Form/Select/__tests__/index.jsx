import {
  render,
  FormikWrapper,
  toMatchSnapshot,
  userEvent,
  screen,
  act,
} from 'test-utils';
import Select from '..';

const cities = [
  { label: 'City 1', value: 1 },
  { label: 'City 2', value: 2 },
];

it('should match snapshot', () => {
  toMatchSnapshot(
    <FormikWrapper initialValues={{ city: '' }}>
      <Select name="city" label="City" options={cities} required />
    </FormikWrapper>,
  );
});

it('should set selected value', async () => {
  render(
    <FormikWrapper initialValues={{ city: '' }}>
      <Select name="city" label="City" options={cities} />
    </FormikWrapper>,
  );

  const select = screen.getByLabelText(/city/i);
  userEvent.click(select);
  await act(async () => {
    userEvent.click(await screen.findByText(cities[0].label));
  });

  expect(select.innerHTML).toBe(cities[0].label);
});

it('should renders empty select with emptyStateMessage', async () => {
  render(
    <FormikWrapper initialValues={{ city: '' }}>
      <Select name="city" label="City" hasEmptyValue />
    </FormikWrapper>,
  );

  const select = screen.getByLabelText(/city/i);
  userEvent.click(select);
  const warnNoOptions = await screen.findByText(/não há opções/i);

  expect(warnNoOptions).toBeInTheDocument();
  expect(warnNoOptions).toHaveAttribute('aria-disabled', 'true');
});

it('should renders empty value', async () => {
  render(
    <FormikWrapper initialValues={{ city: '' }}>
      <Select name="city" label="City" hasEmptyValue options={cities} />
    </FormikWrapper>,
  );

  const select = screen.getByLabelText(/city/i);
  userEvent.click(select);
  const selectOptions = await screen.findAllByRole('option');

  expect(selectOptions).toHaveLength(cities.length + 1);
});

it('should show error', async () => {
  render(
    <FormikWrapper
      initialValues={{ city: '' }}
      validate={() => ({ city: 'Required Field' })}
      onSubmit={() => {}}
    >
      <Select name="city" label="City" options={cities} />
      <button type="submit">submit</button>
    </FormikWrapper>,
  );

  userEvent.click(screen.getByRole('button', { name: /submit/i }));
  expect(await screen.findByText(/required field/i)).toBeInTheDocument();
});
