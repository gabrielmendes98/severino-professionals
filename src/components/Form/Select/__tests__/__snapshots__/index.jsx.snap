// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined Mui-required Mui-required"
        data-shrink="false"
        for="city"
        id="city-label"
      >
        City
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
           *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="city-label city"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
          id="city"
          role="button"
          tabindex="0"
        >
          <span>
            ​
          </span>
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput"
          name="city"
          required=""
          tabindex="-1"
          value=""
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-3"
          >
            <span>
              City *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </form>
  <div
    class="Toastify"
  />
</DocumentFragment>
`;
